agent/acs/handler/heartbeat_handler.go

```diff
@@ -0,0 +1,119 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handler
+
+import (
+	"context"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/cihub/seelog"
+)
+
+// heartbeatHandler handles heartbeat messages from ACS
+type heartbeatHandler struct {
+	heartbeatMessageBuffer    chan *ecsacs.HeartbeatMessage
+	heartbeatAckMessageBuffer chan *ecsacs.HeartbeatAckRequest
+	ctx                       context.Context
+	cancel                    context.CancelFunc
+	acsClient                 wsclient.ClientServer
+}
+
+// newHeartbeatHandler returns an instance of the heartbeatHandler struct
+func newHeartbeatHandler(ctx context.Context,
+	acsClient wsclient.ClientServer) heartbeatHandler {
+
+	// Create a cancelable context from the parent context
+	derivedContext, cancel := context.WithCancel(ctx)
+	return heartbeatHandler{
+		heartbeatMessageBuffer:    make(chan *ecsacs.HeartbeatMessage),
+		heartbeatAckMessageBuffer: make(chan *ecsacs.HeartbeatAckRequest),
+		ctx:                       derivedContext,
+		cancel:                    cancel,
+		acsClient:                 acsClient,
+	}
+}
+
+// handlerFunc returns a function to enqueue requests onto the buffer
+func (heartbeatHandler *heartbeatHandler) handlerFunc() func(message *ecsacs.HeartbeatMessage) {
+	return func(message *ecsacs.HeartbeatMessage) {
+		heartbeatHandler.heartbeatMessageBuffer <- message
+	}
+}
+
+// start() invokes go routines to handle receive and respond to heartbeats
+func (heartbeatHandler *heartbeatHandler) start() {
+	go heartbeatHandler.handleHeartbeatMessage()
+	go heartbeatHandler.sendHeartbeatAck()
+}
+
+func (heartbeatHandler *heartbeatHandler) handleHeartbeatMessage() {
+	for {
+		select {
+		case message := <-heartbeatHandler.heartbeatMessageBuffer:
+			if err := heartbeatHandler.handleSingleHeartbeatMessage(message); err != nil {
+				seelog.Warnf("Unable to handle heartbeat message [%s]: %v", message.String(), err)
+			}
+		case <-heartbeatHandler.ctx.Done():
+			return
+		}
+	}
+}
+
+func (heartbeatHandler *heartbeatHandler) handleSingleHeartbeatMessage(message *ecsacs.HeartbeatMessage) error {
+	seelog.Tracef("Received server heartbeat message: %s", message.MessageId)
```

Comment:
We very rarely use trace logging on our team, so would probably prefer all of these trace log messages changed to `Debugf`

Reply:
Right, I am aware that ECS agent in fact does not really support `TRACE` level logging, so the purpose of this is more towards clarifying the intention, and to show something when running tests.

I'm a bit concerned about the log size if we make them `DEBUG`, as heartbeats will happen *very* often and the likelihood of these logs to be useful is low.

Reply:
same comment as below, probably should change this to `message.String()`

Reply:
> Right, I am aware that ECS agent in fact does not really support `TRACE` level logging, so the purpose of this is more towards clarifying the intention, and to show something when running tests.
> 
> I'm a bit concerned about the log size if we make them `DEBUG`, as heartbeats will happen _very_ often and the likelihood of these logs to be useful is low.

I would say just make them comments then?

Reply:
Cleaned up these trace logging into comments